contracts/delegated_optin_app.algo.ts

```diff
@@ -0,0 +1,186 @@
+
+  // Mapping of hash(sender address + receiver auth address) to
+  // signed address opt-in lsig for sender address
+  addressOptInSignatures = new BoxMap<byte32, byte64>({ prefix: 's-' });
+
+  // Mapping of hash(sender address + receiver address ) to timestamp until which
+  // address pt-ins from the sender address are allowed
+  addressOptInEndTimes = new BoxMap<byte32, uint64>({ prefix: 'e-' });
+
+  private getSenderReceiverHash(sender: Address, receiverOrSigner: Address): byte32 {
+    return sha256(concat(sender, receiverOrSigner));
+  }
+
+  @handle.createApplication
+  create(): void {
+    this.assetMBR.set(100_000);
+  }
+
+  // ************ Meta Methods ************ //
+
+  /**
+   * Set the address of the verifier lsig. This will only be called once after creation.
+   *
+   * @param lsig - The address of the verifier lsig
+   *
+   */
+  setSigVerificationAddress(lsig: Address): void {
+    assert(this.txn.sender === this.app.creator);
+    assert(!this.sigVerificationAddress.exists());
+    this.sigVerificationAddress.set(lsig);
+  }
+
+  /**
+   * Updates the asset MBR
+   *
+   * @param asset - The asset to opt into and opt out of to determine MBR
+   *
+   */
+  updateAssetMBR(asset: Asset): void {
+    const preMbr = this.app.address.minBalance;
+
+    sendAssetTransfer({
+      assetReceiver: this.app.address,
+      xferAsset: asset,
+      assetAmount: 0,
+      fee: 0,
+    });
+
+    const mbrDelta = preMbr - this.app.address.minBalance;
+
+    assert(mbrDelta !== this.assetMBR.get());
+    this.assetMBR.set(mbrDelta);
+
+    sendAssetTransfer({
+      assetReceiver: this.app.address,
+      xferAsset: asset,
+      assetAmount: 0,
+      fee: 0,
+      assetCloseTo: this.app.address,
+    });
+  }
+
+  // ************ Open Opt-In Methods ************ //
+
+  /**
+   * Set the signature of the lsig for the given account
+   *
+   * @param sig - The signature of the lsig
+   * @param signer - The public key corresponding to the signature
+   * @param verifier - A txn from the verifier lsig to verify the signature
+   *
+   */
+  setOpenOptInSignature(sig: byte64, signer: AuthAddr, verifier: Txn): void {
+    assert(verifier.sender === this.sigVerificationAddress.get());
+
+    this.openOptInSignatures.set(signer, sig);
+  }
+
+  /**
+   * Verifies that the opt in is allowed
+   *
+   * @param mbrPayment - Payment to the receiver that covers the ASA MBR
+   * @param optIn - The opt in transaction, presumably from the open opt-in lsig
+   *
+   */
+  openOptIn(mbrPayment: PayTxn, optIn: AssetTransferTxn): void {
+    // Verify mbr payment
+    assert(optIn.assetReceiver === mbrPayment.receiver);
+    assert(mbrPayment.amount >= this.assetMBR.get());
+
+    // If endTimes box exists, openOptIn that the opt in is before the end time
+    if (this.openOptInEndTimes.exists(optIn.assetReceiver)) {
+      assert(this.openOptInEndTimes.get(optIn.assetReceiver) > globals.latestTimestamp);
+    }
```

Comment:
The existance of the signature is not used to determine functionality because I wanted control over functionality to be per-address, rather than per-auth-addr. Rationale explained [here](https://github.com/joe-p/delegated_optin_arc/pull/1#discussion_r1284695447). 

Since a separate box is needed anyways for this, I figured might as well use timestamps to reduce clicks in the event the user knows they only want to delegate for a specific amount of time.

Reply:
What do you mean by per-address, rather than per-auth-addr? Are you talking about the `addressOptIn` path? If so this suggestion applies just as well to that case:

```typescript
assert(this.addressOptInSignatures.exists({ signer: optIn.assetReceiver, sender: this.txn.sender } as SignerAndSender))
```

Reply:
Let's say I have three accounts: account A and account B which both are rekeyed to account C. If I sign the lsig with account C, then anyone can use that signature for account A, B or C. But... maybe I only want to delegate opt-ins for account C, and disable delegated opt-ins on account A and B.

Reply:
Oh I see, you're saying that the key for `openOptInSignatures` represents auth addresses, NOT accounts. So in your example, with accounts A and B rekeyed to C, A and B would both use the signature from `openOptInSignatures[C]`. Meaning we can't use only the existence of boxes from `openOptInSignatures` to indicate opt-in status, since it's a many-to-one mapping, and each account should be able to control that individually.

I have two alternatives suggestions then:
1. Make each `openOptInSignatures` control exactly one account, not an auth address that can back multiple accounts. You'd need to augment the values here, since it's necessary to store the actual signing key for the signature, since it can now be different from the account's key (e.g. vale type could be a tuple of `(bytes32, bytes64)`).
    * Pros:
      * Boxes get cleaned up when they're no longer necessary
    * Cons:
      * Slightly larger box keys to hold info which most of the time won't be necessary (since I assume most accounts aren't rekeyed)
3. Keep the existing convention of `openOptInSignatures`, but instead of having `openOptInEndTimes = new BoxMap<Address, uint64>` control opt-in status, use something like `openOptInEnabled = new BoxMap<Address, null>` (I don't know how you'd represent a zero-length box value in tealscript, but that's what I'm trying to say with `null`). `openOptInEnabled` would control whether an opt-in is allowed for an account: if and only a box exists for that account, it's allowed.
    * Pros:
      * `openOptInEnabled` boxes will get cleaned up when no longer needed.
    * Cons:
      * `openOptInSignatures` boxes won't get cleaned up when no longer needed